### ISSUE TITLE
pyright mixes up types of tuple elements when calling a function with unpacked arguments

### ISSUE THREAD

**Issue by bersbersbers** [microsoft/pyright#5820]
**Describe the bug**
In the code below, `inner(*name_and_num)` looks OK to me - in both cases of `tuple[str] | tuple[str, int]`, its first argument should be `str`. However, `pyright` errors with 

```
C:\Code\project\bug.py:7:12 - error: Argument of type "str | int" cannot be assigned to parameter "_name" of type "str" in function "inner"
  Type "str | int" cannot be assigned to type "str"
    "int" is incompatible with "str" (reportGeneralTypeIssues)
C:\Code\project\bug.py:7:12 - error: Argument of type "str | int" cannot be assigned to parameter "_num" of type "int" in function "inner"
  Type "str | int" cannot be assigned to type "int"
    "str" is incompatible with "int" (reportGeneralTypeIssues)
```

Some similar confusion seems to exist for the second argument. 

**Code or Screenshots**
```python
def inner(_name: str, _num: int = 0) -> None:
    ...

def outer(name_and_num: tuple[str] | tuple[str, int]) -> None:
    inner(name_and_num[0])  # passes type checks
    inner(*name_and_num)  # does not pass type-checks: "int" is incompatible with "str"

outer(("",))
outer(("", 0))
```

**VS Code extension or command-line**
`pyright` 1.1.324 from the command line

**Comment by erictraut** [comment]
This is working as intended currently, so I don't consider this a bug. I'll re-label it a feature enhancement request.

Currently, when pyright encounters an unpacked argument that is a union of tuples, it "merges" the tuples into a single tuple and uses that combined type to validate the call expression. In this case, the two tuples are different lengths, so they get merged into the tuple `tuple[str | int, ...]`. That explains the error message you're seeing here.

It is theoretically possible to do "union expansion" and type check the call with all combinations of expanded unions. Pyright does this for overloads (not because I think it's a good idea, but because mypy chose to do it, and I felt compelled to mirror mypy's behavior for compatibility reasons). Union expansion is very costly at analysis time, so doing it for every call (even non-overloads) is problematic. This would address a few edge cases like the one in your code sample above, but it would measurably slow down type analysis for all code. For that reason, its a feature that would have net negative value to pyright users. For that reason, I'm going to reject the feature request and close it for now. This is a decision we could revisit in the future if we receive sufficient signal from pyright users that these are more than just rare edge cases.

In the meantime, there are straightforward ways to work around this limitation in your code and avoid the edge case.

**Comment by bersbersbers** [comment]
Thanks!

> In the meantime, there are straightforward ways to work around this limitation in your code and avoid the edge case.

It gets less straightforward if the number of tuples exceeds 2, but the code is not super-readable in that case anyway, so I went ahead defining a `dataclass` for this. Thanks again!
